src/graph/TruncateSpaceExecutor.cpp

```diff
@@ -0,0 +1,199 @@
+                                         ExecutionContext *ectx)
+        : Executor(ectx) {
+    sentence_ = static_cast<TruncateSpaceSentence*>(sentence);
+}
+
+Status TruncateSpaceExecutor::prepare() {
+    spaceName_ = sentence_->spaceName();
+    return Status::OK();
+}
+
+void TruncateSpaceExecutor::truncateSpace() {
+    auto future = ectx()->getMetaClient()->getSpace(*spaceName_);
+    auto *runner = ectx()->rctx()->runner();
+
+    auto cb = [this] (auto &&resp) {
+        if (!resp.ok()) {
+            onError_(Status::Error("Space not found"));
+            return;
+        }
+
+        fromSpaceId_ = resp.value().get_space_id();
+        tempSpaceName_ = folly::stringPrintf("temp_to_space_%ld",
+                                            time::WallClock::fastNowInMicroSec());
+        auto properties = resp.value().get_properties();
+        meta::SpaceDesc spaceDesc = meta::SpaceDesc(tempSpaceName_,
+                                                    properties.get_partition_num(),
+                                                    properties.get_replica_factor(),
+                                                    properties.get_charset_name(),
+                                                    properties.get_collate_name());
+        createSpace(spaceDesc);
+    };
+
+    auto error = [this] (auto &&e) {
+        LOG(ERROR) << "Exception caught: " << e.what();
+        onError_(Status::Error("Describe space exception: %s", e.what().c_str()));
+        return;
+    };
+
+    std::move(future).via(runner).thenValue(cb).thenError(error);
+}
+
+void TruncateSpaceExecutor::createSpace(const meta::SpaceDesc &spaceDesc) {
+    auto future = ectx()->getMetaClient()->createSpace(spaceDesc, false);
+    auto *runner = ectx()->rctx()->runner();
+
+    auto cb = [this] (auto &&resp) {
+        if (!resp.ok()) {
+            onError_(Status::Error("Create space `%s' failed: %s.",
+                                   tempSpaceName_.c_str(), resp.status().toString().c_str()));
+            return;
+        }
+        newSpaceId_ = std::move(resp).value();
+        if (newSpaceId_ <= 0) {
+            onError_(Status::Error("Create space failed"));
+            return;
+        }
+        copySpace();
+    };
+
+    auto error = [this] (auto &&e) {
+        auto msg = folly::stringPrintf("Create space `%s' exception: %s.",
+                                        spaceName_->c_str(), e.what().c_str());
+        LOG(ERROR) << msg;
+        onError_(Status::Error(std::move(msg)));
+        return;
+    };
+
+    std::move(future).via(runner).thenValue(cb).thenError(error);
+}
+
+void TruncateSpaceExecutor::copySpace() {
+    auto future = ectx()->getMetaClient()->copySchemaFromSpace(
+            tempSpaceName_, *spaceName_, true);
+    auto *runner = ectx()->rctx()->runner();
+
+    auto cb = [this] (auto &&resp) {
+        if (!resp.ok()) {
+            onError_(Status::Error("Copy space `%s' failed: %s.",
```

Comment:
need delete the temp space when copy failed?

Reply:
good catch